File: go/simple-linked-list/simple_linked_list.go
package linkedlist

import (
	"errors"
)

// Element represents a node in the linked list
type Element struct {
	value int
	next  *Element
}

// List represents a singly linked list with a head pointer and size
type List struct {
	head *Element
	size int
}

// New creates a new linked list from a slice of integers
func New(elements []int) *List {
	list := &List{}
	
	if elements == nil || len(elements) == 0 {
		return list
	}
	
	// Add each element to the list
	for _, val := range elements {
		list.Push(val)
	}
	
	return list
}

// Size returns the number of elements in the list
func (l *List) Size() int {
	return l.size
}

// Push adds an element to the end of the list
func (l *List) Push(element int) {
	newElement := &Element{
		value: element,
		next:  nil,
	}
	
	// If the list is empty, set the head to the new element
	if l.head == nil {
		l.head = newElement
	} else {
		// Otherwise, traverse to the end of the list and add the new element
		current := l.head
		for current.next != nil {
			current = current.next
		}
		current.next = newElement
	}
	
	l.size++
}

// Pop removes and returns the last element from the list
func (l *List) Pop() (int, error) {
	if l.head == nil {
		return 0, errors.New("list is empty")
	}
	
	// If there's only one element, remove it and return its value
	if l.head.next == nil {
		value := l.head.value
		l.head = nil
		l.size--
		return value, nil
	}
	
	// Otherwise, traverse to the second-to-last element
	current := l.head
	for current.next.next != nil {
		current = current.next
	}
	
	// Get the value of the last element
	value := current.next.value
	
	// Remove the last element
	current.next = nil
	l.size--
	
	return value, nil
}

// Array converts the list to a slice of integers
func (l *List) Array() []int {
	if l.head == nil {
		return []int{}
	}
	
	result := make([]int, 0, l.size)
	current := l.head
	
	// Traverse the list and add each value to the result
	for current != nil {
		result = append(result, current.value)
		current = current.next
	}
	
	return result
}

// Reverse returns a new list with the elements in reverse order
func (l *List) Reverse() *List {
	if l.head == nil {
		return New([]int{})
	}
	
	// Convert the list to an array
	arr := l.Array()
	
	// Reverse the array
	for i, j := 0, len(arr)-1; i < j; i, j = i+1, j-1 {
		arr[i], arr[j] = arr[j], arr[i]
	}
	
	// Create a new list from the reversed array
	return New(arr)
}

File: python/simple-linked-list/simple_linked_list.py
class EmptyListException(Exception):
    """Exception raised when trying to access elements of an empty list."""
    pass


class Node:
    def __init__(self, value, next_node=None):
        self._value = value
        self._next = next_node

    def value(self):
        return self._value

    def next(self):
        return self._next


class LinkedList:
    def __init__(self, values=None):
        self._head = None
        self._length = 0
        
        if values is not None:
            for value in values:
                self.push(value)

    def __iter__(self):
        current = self._head
        while current is not None:
            yield current.value()
            current = current.next()

    def __len__(self):
        return self._length

    def head(self):
        if self._head is None:
            raise EmptyListException("The list is empty.")
        return self._head

    def push(self, value):
        self._head = Node(value, self._head)
        self._length += 1

    def pop(self):
        if self._head is None:
            raise EmptyListException("The list is empty.")
        
        value = self._head.value()
        self._head = self._head.next()
        self._length -= 1
        
        return value

    def reversed(self):
        result = LinkedList()
        values = []
        
        # Collect values in the current order
        current = self._head
        while current is not None:
            values.append(current.value())
            current = current.next()
        
        # Add them in reverse order to get the original order when iterated
        for value in values:
            result.push(value)
            
        return result
